### PR TITLE
WIP: Add 'interfaces' service template

### PR DESCRIPTION
**Pull request by ndom91**
Add `interfaces` service template to ease the pulling of interface details with the end goal of creating a csv which one can easily import into the [Netbox](https://github.com/netbox-community/netbox) bulk interface importer. 

Current idea is to start with very basics and maybe just wrap this in a little CLI which can simply output the csv which one can paste into netbox. Further enhancements could include directly calling the netbox api to add said interfaces or something like that.

#### Notes

Netbox interface importer example: https://netboxdemo.com/dcim/interfaces/import
> Login: **netbox**/**netbox**

Fields:
```
device, name, type, enabled, mac_address, mtu, description, mode 
```
- device - device name which the interface belongs to
- name - interface name
- type - physical medium (SFP, 10/100, 1G, etc.)
- enabled - boolean on/off
- mac_address - mac address
- mtu - mtu size
- description - description
- mode - 802.1Q Mode (access, tagged, tagged (all))

### CODE REVIEW

File: backend/plugins/extensibles/j2_service_templates/interfaces.j2 [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+						"username": "{{username}}",
+						"password": "{{password}}"
+					},
+					"command": "{{command}}",
+					"args":{
+						"use_textfsm": "true"
+					}
+        }
+      },
+      {
```
Comment (tbotnz) [inline backend/plugins/extensibles/j2_service_templates/interfaces.j2:54]: please remove this object as its not required

File: backend/plugins/extensibles/j2_service_templates/interfaces.j2 [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		"hosts":["10.0.2.25","10.0.2.23"],
+	}
+}
+!!!! ANY PAYLOAD CAN BE ADDED TO THE ARGS DICT, THESE ARE THEN AVAILABLE IN THE J2 TEMPLATE AS BELOW
+#}
+[
+  {% for host in hosts %}
+  {
+    "supported_methods": [
+      {
```
Comment (tbotnz) [inline backend/plugins/extensibles/j2_service_templates/interfaces.j2:30]: please remove this object as its not required

File: backend/plugins/extensibles/j2_service_templates/interfaces_model.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    operation: service_lifecycle
+    args: interface_model_args
+    queue_strategy: Optional[queue_strat] = None
+
+    class Config:
+        # add an example payload under the "example" dict 
+        schema_extra = {
+            "example": {
+                "operation": "retrieve",
+                "args":{
```
Comment (tbotnz) [inline backend/plugins/extensibles/j2_service_templates/interfaces_model.py:26]: the args should also include driver & command

Comment (ndom91) [reply backend/plugins/extensibles/j2_service_templates/interfaces_model.py]: Whoops forgot to commit that update. Its in now.

**Comment by ndom91** [comment]
Updated the name and cleaned up the unnecessary objects and what not.

I accidentally commited my local copy of the docker-compose and config.json files accidentally and then deleted them. I hope merging this wont delete them completely. If so, I'll have to see if I can revert the last two commits and redo that correctly :sweat_smile:

File: backend/plugins/extensibles/j2_service_templates/netmiko_retrieve_data_model.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+class interface_model_args(BaseModel):
+    # your model goes here!
+    hosts: list
+    username: str
+    password: str
+    driver: str
+    command: str
+
+class interface_model(BaseModel):
```
Comment (tbotnz) [inline backend/plugins/extensibles/j2_service_templates/netmiko_retrieve_data_model.py:15]: plz rename this netmiko_retrieve_data_model

File: backend/plugins/extensibles/j2_service_templates/netmiko_retrieve_data_model.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+from pydantic import BaseModel
+from typing import Dict, FrozenSet, List, Optional, Sequence, Set, Tuple, Union
+
+from backend.core.models.service import service_lifecycle
+from backend.core.models.models import queue_strat
+
+class interface_model_args(BaseModel):
```
Comment (tbotnz) [inline backend/plugins/extensibles/j2_service_templates/netmiko_retrieve_data_model.py:7]: plz rename this netmiko_retrieve_data_model_args

File: backend/plugins/extensibles/j2_service_templates/netmiko_retrieve_data_model.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    hosts: list
+    username: str
+    password: str
+    driver: str
+    command: str
+
+class interface_model(BaseModel):
+    # this class MUST match the filename & the filename must be formatted $servicetemplatename_model.py
+    operation: service_lifecycle
+    args: interface_model_args
```
Comment (tbotnz) [inline backend/plugins/extensibles/j2_service_templates/netmiko_retrieve_data_model.py:18]: plz rename this netmiko_retrieve_data_model_args

File: backend/plugins/extensibles/j2_service_templates/netmiko_retrieve_data.j2 [hunk 7]
```diff
@@ -0,0 +1,39 @@
+{# NETPALM SERVICE BOILERPLATE #}
```
Comment (tbotnz) [inline backend/plugins/extensibles/j2_service_templates/netmiko_retrieve_data.j2:1]: the comments can be removed {# #}

**Comment by ndom91** [comment]
Okay, should be cleaned up then :+1:

**Comment by ndom91** [comment]
For some reason I can't get this to work in the web openapi docs although it was working fine in my postman setup in the office. 

I now get the error:
```
 "validate_template Expecting property name enclosed in double quotes: line 13 column 6 (char 276)",
      "render_service 'NoneType' object is not iterable"
```
returned when I check the task

File: config.json [hunk 8]
```diff
@@ -1,40 +0,0 @@
-{
```
Comment (tbotnz) [inline config.json:1]: why remove?

File: docker-compose.yml [hunk 9]
```diff
@@ -1,56 +0,0 @@
-version: "3.7"
```
Comment (tbotnz) [inline docker-compose.yml:1]: why remove?

Comment (ndom91) [reply docker-compose.yml:1]: Yeah didn't mean to. I had accidentally commit my local copies that had the volumes mounted, in docker-compose and my own api-key in config.json, then when I tried to remove my accidental commit from this PR it lead to this mess.. Haha.

Comment (tbotnz) [reply docker-compose.yml:1]: No worries, just delete the two files from your PR and ill merge it

Comment (ndom91) [reply docker-compose.yml]: Okay so I restored those two files back to the state of the upstream that I forked from, so I'm not sure why theres those few small differences in there now.

**Pull request closed without merging**